### PR TITLE
Earth - Sophie - Media Ranker

### PR DESCRIPTION
**Pull request by steve-messing**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | My top_10 method in work.rb uses helper methods to filter by category, and then sorts each list by number of votes and secondarily by alphabetical order to handle ties.
Describe how you approached testing that model method. What edge cases did you come up with? | I checked that the top ten list could be shorter than 10 if less than 10 books exist in the database. I also checked that the book with title "AAAA" with 0 votes would occur before book title "MMMM" with 0 votes.
What are `session` and `flash`? What is the difference between them? | session tracks the current user logged in. Flash is a hash containing error and success messages, that are flashed to the user when they take certain actions in the app.
What was one thing that you gained more clarity on through this assignment? | Indirect relationships and db structures
What is the Heroku URL of your deployed application? | https://sm-media-ranker.herokuapp.com/

### CODE REVIEW

File: app/views/users/show.html.erb [hunk 1]
```diff
@@ -0,0 +1,27 @@
+<h4><strong>User: </strong><%= @user.username %></h4>
+<p><%= @user.votes.count %> Votes</p>
+<p>Active since <%= @user.created_at.to_date %></p>
+
+<table class="table">
```
Comment (CheezItMan) [inline app/views/users/show.html.erb:5]: Notice you make this kind of table, a **lot**, maybe you could use a partial!  Hmmm... maybe the partial you have below?

File: app/models/work.rb [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    categories(:book)
+  end
+
+  def self.albums
+    categories(:album)
+  end
+
+
+  def self.top_10(media)
+    media_list = categories(media).sort_by { |work| [-work.votes.count, work.title] }[0..9]
```
Comment (CheezItMan) [inline app/models/work.rb:25]: This works, but I strongly suggest in the future finding ways to use `.order` and `.where` and `.limit` to let Postgres do the sorting and filtering.  That way it doesn't have to transmit **every** work to rails to then sort and filter.  

Postgres is very good at doing this sort of thing with large amounts of data, let it. 

[Here is an article on it](https://tomkadwill.com/2017/12/18/counting-or-sorting-by-an-association-in-a-rails-sql-query.html)

File: app/models/work.rb [hunk 3]
```diff
@@ -0,0 +1,44 @@
+class Work < ApplicationRecord
```
Comment (CheezItMan) [inline app/models/work.rb:1]: I like these helper methods and validations

File: app/controllers/works_controller.rb [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    @works = Work.all
+    @books = Work.books
+    @movies = Work.movies
+    @albums = Work.albums
+  end
+
+  def show
+    if @work.nil?
+      head :not_found
+      return
```
Comment (CheezItMan) [inline app/controllers/works_controller.rb:23]: At this point just redirect them with a flash notice

File: app/controllers/works_controller.rb [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+  def create
+    @work = Work.new(work_params)
+
+    if @work.save
+      flash[:success] = "#{@work.title} was successfully added!"
+      redirect_to work_path(@work)
+      return
+    else
+      flash.now[:failure] = "#{@work.title} was not successfully added."
```
Comment (CheezItMan) [inline app/controllers/works_controller.rb:39]: You should also provide the validation errors.

File: app/controllers/works_controller.rb [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+      render :new, status: :bad_request
+      return
+    end
+  end
+
+  def edit
+    if @work.nil?
+      redirect_to works_path
```
Comment (CheezItMan) [inline app/controllers/works_controller.rb:47]:
```suggestion
      flash[:notice] = "Unable to find work #{params[:id]}"
      redirect_to works_path
```

File: app/controllers/works_controller.rb [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+  def update
+    if @work.nil?
+      head :not_found
+      return
+    elsif @work.update(work_params)
+      flash[:success] = "#{@work.category.capitalize} was successfully updated!"
+      redirect_to work_path(@work)
+      return
+    else
+      flash.now[:failure] = "#{@work.title} was not successfully updated."
```
Comment (CheezItMan) [inline app/controllers/works_controller.rb:61]: You should tell them why with the validation errors.

File: app/controllers/works_controller.rb [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+      render :edit, status: :bad_request
+      return
+    end
+  end
+
+  def destroy
+    if @work.nil?
+      head :not_found
+      return
```
Comment (CheezItMan) [inline app/controllers/works_controller.rb:70]: Just redirect them with a flash notice

File: app/controllers/works_controller.rb [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+    else
+      @work.destroy
+      redirect_to works_path
```
Comment (CheezItMan) [inline app/controllers/works_controller.rb:73]: Maybe include a flash notice to tell them it worked?

File: test/models/user_test.rb [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+    @movie6 = works(:movie6)
+    @movie7 = works(:movie7)
+    @movie8 = works(:movie8)
+    @movie9 = works(:movie9)
+    @movie10 = works(:movie10)
+    @movie11 = works(:movie11)
+    @movie12 = works(:movie12)
+  end
+
+  describe "validations" do
```
Comment (CheezItMan) [inline test/models/user_test.rb:29]: Also check the error message on validations

File: test/models/user_test.rb [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+      expect(result).must_equal false
+      expect(@user.errors).must_include :username
+    end
+  end
+
+  describe "relationships" do
+
+    it "user can have many votes" do
+      expect(@user1.votes.count).must_equal 7
+    end
```
Comment (CheezItMan) [inline test/models/user_test.rb:60]: Also check that each element of `@user1.vote` is an instance of `Vote`

File: test/models/vote_test.rb [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+    @vote1 = votes(:vote1)
+    @vote2 = votes(:vote2)
+    @vote3 = votes(:vote3)
+
+    @user1 = users(:user1)
+
+    @work1 = works(:movie1)
+  end
+
+  describe "relationships" do
```
Comment (CheezItMan) [inline test/models/vote_test.rb:15]: These are good, just also test the error message.

File: test/models/work_test.rb [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+      )
+      result = @new_movie.valid?
+      expect(result).must_equal false
+    end
+
+  end
+
+  describe "categories" do
+
+    it "can produce a list of media for each category" do
```
Comment (CheezItMan) [inline test/models/work_test.rb:71]: Also check that it works if there are no works in the category.

File: test/models/work_test.rb [hunk 14]
```diff
@@ -0,0 +1,146 @@
...
+
+      expect(@top_movies[4].title).must_equal "AAAAAA"
+      expect(@top_movies[4].votes.count).must_equal 0
+
+      expect(@top_movies.last.title).must_equal "MMMM"
+      expect(@top_movies.last.votes.count).must_equal 0
+    end
+  end
+
+  describe "spotlight" do
```
Comment (CheezItMan) [inline test/models/work_test.rb:140]: What about when there are no works?

**Review by CheezItMan — COMMENTED** [review]
# Media Ranker





## Functional Requirements: Manual Testing

| Criteria | yes/no |
| --- | --- |
| **Before logging in** | --
| 1. On index page, there are at most 10 pieces of media on three lists, and a Media Spotlight | ✔️
| 2. Can go into a work's show page | ✔️
| 3. Verify unable to vote on a work, and get a flash message | ✔️
| 4. Can edit this work successfully, and get a flash message | ✔️, it makes me log in 1st, neat 💯
| 5. Can go to "View all media" page and see three lists of works, sorted by vote | ⚠️, they're not sorted by vote
| 6. Verify unable to create a new work when the form is empty, and details about the validation errors are visible to the user through a flash message | ✔️
| 7. Can create a new work successfully. Note the URL for this work's show page | ✔️
| 8. Can delete this work successfully |✔️
| 9. Going back to the URL of this deleted work's show page produces a 404 or some redirect behavior (and does not try to produce a broken view) | ✔️, it produces a 404 with the `head` command, I really recommend a redirect instead.
| 10. Verify that the "View all users" page lists no users | ✔️
| **Log in** | --
| 11. Logging in with a valid name changes the UI to "Logged in as" and "Logout" buttons | ✔️
| 12. Your username is listed in "View all users" page | ✔️
| 13. Verify that number of votes determines the Media Spotlight | ✔️
| 14. Voting on several different pieces of media affects the "Votes" tables shown in the work's show page and the user's show page | ✔️
| 15. Voting on the same work twice produces an error and flash message, and there is no extra vote | ✔️
| **Log out** | --
| 16. Logging out showed a flash message and changed the UI | ✔️
| 17. Logging in as a new user creates a new user | ✔️
| 18. Logging in as an already existing user has a specific flash message | ✔️

## Major Learning Goals/Code Review

| Criteria | yes/no |
| --- | --- |
| 1. Sees the full development cycle including deployment, and the app is deployed to Heroku | ✔️
| 2. Practices full-stack development and fulfilling story requirements: the styling, look, and feel of the app is similar to the original Media Ranker | ✔️, a little different, but close'nuf
| 3. Practices git with at least 25 small commits and meaningful commit messages | ⚠️, not a lot of commits 

### Previous Rails learning, Building Complex Model Logic, DRYing up Rails Code

| Criteria | yes/no |
| --- | --- |
| 4. Routes file uses `resources` for works | ✔️?
| 5. Routes file shows intention in limiting routes for voting, log-in functionality, and users | ✔️
| 6. The homepage view, all media view, and new works view use semantic HTML | ✔️
| 7. The homepage view, all media view, and new works view use partials when appropriate | ✔️, **awesome**
| 8. The model for media (likely named `work.rb`) `has_many` votes | ✔️
| 9. The model for media has methods to describe business logic, specifically for top ten and top media, possibly also for getting works by some category | ✔️
| 10. Some controller, likely the `ApplicationController`, has a controller filter for finding a logged in user | ✔️
| 11. Some controller, likely the `WorksController`, has a controller filter for finding a work | ✔️, 💯
| 12. The `WorksController` uses strong params | ✔️
| 13. The `WorksController`'s code style is clean, and focused on working with requests, responses, `params`, `session`, `flash` | ⚠️, some improvements to be done.

### Testing Rails Apps

| Criteria | yes/no |
| --- | --- |
| 14. There are valid fixtures files used for users, votes, and works | ✔️
| 15. User model has tests with sections on validations (valid and invalid) and relationships (has votes) | ✔️
| 16. Vote model has tests with sections on validations (valid and invalid) and relationships (belongs to a user, belongs to a vote) | ✔️
| 17. Work model has tests with sections on validations (valid and invalid) and relationships (has votes) | ✔️
| 18. Work model has tests with a section on all business logic methods in the model, including their edge cases | ✔️, yes but check the edge cases, overall very good though

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 14+ in Functional Requirements: Manual Testing && 14+ in Code Review | 💚  






## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Perfect Indentation | ✅
| Elegant/Clever | ✅
| Descriptive/Readable | ✅
| Concise | ✅
| Logical/Organized | ✅

## Summary

This is really great work Sophie.  You hit the learning goals.  I really like the site.  

Some prioritized areas for improvement:

1.  Don't just use `head` it's ugly for human users to see. 
1.  Look for edge-cases in tests
1.  Work to include validation error messages with flash notices
1.  Take a look at my note about using Postgres to sort and filter things, it's **much** more scalable.  This is a minor note.
